### DISCUSSION TITLE
4 settembre 2023 - esercizio 3

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/basi-di-dati-1#46] [Esami]
E' dato un file di 145.950 record. Ogni record occupa 255 byte. Un blocco contiene 2048 byte.

Un puntatore a blocco occupa 5 byte. Si utilizza una organizzazione hash con record distribuiti uniformemente tra 250 bucket.

3a) Calcolare l'occupazione in blocchi della bucket directory e dei bucket

3b) Calcolare il costo medio di una ricerca considerando chiavi di hash univoche

3c) Quanti bucket occorrerebbero per avere un costo medio di ricerca minore o uguale a 10 accessi?

**Comment by Warcophyr** [comment]
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/133346488/d9b0a487-899f-400e-a125-ea4af3f7fe23)

**Reply by niccolo-fato** [reply]
Teoricamente a NBLKxBCK non dovevi fare 584/8?

**Reply by FeddyLix17** [reply]
- per il costo medio non hai incluso l'accesso (singolo, dato che contiene un solo blocco), alla bucket directory

- come notificato anche da @niccolo-fato , nonostante l'averlo calcolato bene inizialmente, hai considerato come se un blocco potesse contenere 5 record (invece di 8)

- la disuguaglianza del punto 3c) di per se è sviluppata correttamente, ma le sviste nei 2 punti precedenti portano a un risultato leggermente diverso da quello corretto

**Comment by niccolo-fato** [comment]
![IMAGE 2024-01-14 08:19:30](https://github.com/sapienzastudentsnetwork/bdi2324/assets/73995334/21406ca0-4971-400e-b5ef-78aafabd35d4)

**Comment by daromacarlo** [comment]
![Immagine 2024-01-23 160439](https://github.com/sapienzastudentsnetwork/bdi2324/assets/135441131/1b780c4b-d5ce-47ca-80b9-1a60e074951b)
![Immagine 2024-01-23 160449](https://github.com/sapienzastudentsnetwork/bdi2324/assets/135441131/e4212872-f452-4eb1-8780-f8923b705635)

**Comment by FeddyLix17** [comment]
3a)

Avendo i record distribuiti uniformemente tra <ins>250 bucket</ins>, avremo, di conseguenza, 250 puntatori (ognuno per il corrispettivo bucket), presenti nella bucket directory, i quali occuperanno complessivamente $\Large 250 \times 5 = 1250\ byte$, dunque sarà necessario 

$$\Large n^{\circ}\ Blocchi\ Bucket\ Directory = \left\lceil \frac{n^{\circ}\ Puntatori \times Pointer\ Size}{Block\ Size} \right\rceil = 1$$

blocco per la bucket directory.

Abbiamo 145.950 record da immagazzinare <ins>uniformemente</ins> in 250 bucket, dunque

$$\Large Record\ per\ Bucket = \left\lceil \frac{n^{\circ}\ Record}{n^{\circ}\ Bucket} \right\rceil = \left\lceil \frac{145.950}{250} \right\rceil = \left\lceil 583,8 \right\rceil = 584$$

$$\Large Record\ per\ Blocco = \left\lfloor \frac{Block\ Size}{Record\ Size} \right\rfloor = \left\lfloor \frac{2048}{255} \right\rfloor = \left\lfloor 8,03 \right\rfloor = 8$$

$$\Large Blocchi\ per\ Bucket = \left\lceil \frac{Record\ per\ Bucket}{Record\ per\ Blocco} \right\rceil = \left\lceil \frac{584}{8} \right\rceil = \left\lceil 73 \right\rceil = 73$$

per un totale di $\Large 73 \times 250 = 18.250$ blocchi necessari per i bucket. 

---

3b)

Al singolo accesso al singolo blocco della bucket directory bisognerà aggiungere i

$$\Large \left\lceil \frac{73}{2} \right\rceil = \left\lceil 36,5 \right\rceil  = 37$$

per il costo medio di una ricerca con chiavi univoche (l'unicità della chiave ci garantisce l'esistenza del record)

> un record si troverà [equiprobabilmente](https://en.wikipedia.org/wiki/Discrete_uniform_distribution) in ognuno dei blocchi di quel particolare bucket

---

3c)

Definiamo $\Large Bk$ come il numero di bucket desiderato

$$\Large Record\ per\ bucket\ Bk = \frac{145.950}{Bk}$$

$$\Large Blocchi\ per\ bucket\ Bk = \left\lceil \frac{145.950}{Bk \times 8} \right\rceil$$

$$\Large \underbrace{1 + \left\lceil\frac{Blocchi\ per\ bucket\ Bk}{2} \right\rceil}_{costo\ medio \atop di\ ricerca} \leq 10$$

Il massimo numero di blocchi che ogni Bucket (di tutti i $\Large Bk$) sarà

$$\Large 1 + \left\lceil\frac{Blocchi\ per\ bucket\ Bk}{2} \right\rceil \leq 10$$

$$\Large \left\lceil\frac{Blocchi\ per\ bucket\ Bk}{2} \right\rceil \leq 9$$

$$\Large \left\lceil Blocchi\ per\ bucket\ Bk\right\rceil \leq 18$$

Da cui l'effettivo numero di Bucket $\Large Bk$

$$\Large \left\lceil \frac{145.950}{Bk \times 8} \right\rceil \leq 18$$

$$\Large \left\lceil \frac{145.950}{8} \right\rceil \leq 18Bk$$

$$\Large Bk \geq \left\lceil\frac{145.950}{18 \times 8}\right\rceil$$

$$\Large Bk \geq \left\lceil 1013,54 \right\rceil$$

$$\Large Bk \geq 1014$$


> Verifica con  $\Large BK = 1014$
>
> $$\Large Record\ per\ bucket: \left\lceil\frac{145.950}{1014}\right\rceil = \left\lceil143,97\right\rceil = 144$$
>
> $$\Large Blocchi\ per\ bucket =  \left\lceil \frac{144}{8} \right\rceil = \left\lceil 18 \right\rceil = 18$$
>
> $$\Large Costo\ medio = 1 + \left\lceil\frac{18}{2}\right\rceil = 10\ accessi$$


> Verifica con  $\Large BK = 1013$ (valida anche per ogni altro valore minore)
>
> $$\Large Record\ per\ bucket: \left\lceil\frac{145.950}{1013}\right\rceil = \left\lceil144,05\right\rceil = 145$$
>
>
> $$\Large Blocchi\ per\ bucket =  \left\lceil \frac{145}{8} \right\rceil = \left\lceil 18,12 \right\rceil = 19$$
>
> $$\Large Costo\ medio = 1 + \left\lceil\frac{19}{2}\right\rceil = \left\lceil10,5\right\rceil = 11\ accessi$$
